db/migrate/20150213162341_add_value_constraints_batch_1.rb

```diff
@@ -0,0 +1,1311 @@
+
+-- see count of NULLs in proposed key columns:
+--   SELECT COUNT(*) AS "total number of rows", SUM((site_id IS NULL)::int) AS "site_id NULLs", SUM((specie_id IS NULL)::int) AS "species_id NULLs", SUM((citation_id IS NULL)::int) AS "citation_id NULLs", SUM((cultivar_id IS NULL)::int) AS "cultivar_id NULLs", SUM((treatment_id IS NULL)::int) AS "treatment_id NULLs", SUM((date IS NULL)::int) AS "date NULLs", SUM((time IS NULL)::int) AS "time NULLs", SUM((variable_id IS NULL)::int) AS "variable_id NULLs", SUM((entity_id IS NULL)::int) AS "entity_id NULLs", SUM((method_id IS NULL)::int) AS "method_id NULLs", SUM((date_year IS NULL)::int) AS "date_year NULLs", SUM((date_month IS NULL)::int) AS "date_month NULLs", SUM((date_day IS NULL)::int) AS "date_day NULLs", SUM((time_hour IS NULL)::int) AS "time_hour NULLs", SUM((time_minute)::int) AS "time_minute" FROM traits;
+-- more info about missing date and time info:
+--   SELECT COUNT(*) AS "total number of rows", SUM((date IS NULL AND date_year IS NULL AND date_month IS NULL AND date_day IS NULL)::int) AS "rows with no date info", SUM((time IS NULL AND time_hour IS NULL AND time_minute IS NULL)::int) AS "rows with no time into" FROM traits;
+
+
+/* TREATMENTS */
+
+ALTER TABLE treatments ALTER COLUMN name SET NOT NULL;
+/* FIX */ --UPDATE treatments SET name = normalize_whitespace(name) WHERE NOT is_whitespace_normalized(name);
+ALTER TABLE treatments ADD CONSTRAINT normalized_treatment_name CHECK (is_whitespace_normalized(name));
+ALTER TABLE treatments ALTER COLUMN definition SET NOT NULL,
+                       ALTER COLUMN definition SET DEFAULT '',
+                       ADD CONSTRAINT normalized_treatment_definition CHECK (is_whitespace_normalized(definition));
+
+
+/* USERS */
+
+ALTER TABLE users ALTER COLUMN login SET NOT NULL;
+ALTER TABLE users ADD CONSTRAINT valid_user_login CHECK (login ~ '^[a-z\d_][a-z\d_.@-]{2,39}$'); -- matches Rails app requirements
+ALTER TABLE users ALTER COLUMN name SET NOT NULL,
+                  ALTER COLUMN name SET DEFAULT '';
+ALTER TABLE users ADD CONSTRAINT normalized_user_name CHECK (is_whitespace_normalized(name));
+ALTER TABLE users ALTER COLUMN email SET NOT NULL;
+ALTER TABLE users ADD CONSTRAINT well_formed_user_email CHECK (is_wellformed_email(email));
+ALTER TABLE users ALTER COLUMN city SET NOT NULL,
+                  ALTER COLUMN city SET DEFAULT '';
+/* FIX */ --UPDATE users SET city = normalize_whitespace(city) WHERE NOT is_whitespace_normalized(city);
+ALTER TABLE users ADD CONSTRAINT normalized_user_city_name CHECK (is_whitespace_normalized(city));
+ALTER TABLE users ALTER COLUMN country SET NOT NULL,
+                  ALTER COLUMN country SET DEFAULT '';
+ALTER TABLE users ADD CONSTRAINT normalized_user_country_name CHECK (is_whitespace_normalized(country));
+ALTER TABLE users ALTER COLUMN crypted_password SET NOT NULL;
+ALTER TABLE users ADD CONSTRAINT valid_user_crypted_password_value CHECK (crypted_password ~ '^[0-9a-f]{40}$');
```

Comment:
I thought the plan was to drop the 40 chars? the php code I use does not have 40 chars password generated. I think it might depend on the salt.